kernel/examples/write-table/src/main.rs

```diff
@@ -0,0 +1,322 @@
+use arrow::util::pretty::print_batches;
+use clap::Args;
+use common::LocationArgs;
+
+use delta_kernel::arrow::array::TimestampMicrosecondArray;
+use delta_kernel::engine::arrow_conversion::TryIntoArrow;
+use delta_kernel::engine::arrow_data::ArrowEngineData;
+use delta_kernel::engine::default::executor::tokio::TokioBackgroundExecutor;
+use delta_kernel::engine::default::DefaultEngine;
+use delta_kernel::object_store::{path::Path, ObjectStore};
+use delta_kernel::schema::{DataType, SchemaRef, StructField, StructType};
+use delta_kernel::{DeltaResult, Engine, Snapshot};
+use serde_json::{json, to_vec};
+use url::Url;
+
+use clap::Parser;
+use itertools::Itertools;
+
+/// An example program that writes to a Delta table and creates it if necessary.
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+#[command(propagate_version = true)]
+struct Cli {
+    #[command(flatten)]
+    write_args: WriteArgs,
+
+    #[command(flatten)]
+    location_args: LocationArgs,
+}
+
+#[derive(Args)]
+pub struct WriteArgs {
+    /// Comma-separated schema specification of the form `field_name:data_type`
+    #[arg(
+        long,
+        short = 's',
+        default_value = "id:integer,name:string,score:double"
+    )]
+    schema: String,
+
+    /// Number of rows to generate for the example data
+    #[arg(long, short, default_value = "10")]
+    num_rows: usize,
+    // TODO: Support providing input data from a JSON file instead of generating random data
+    // TODO: Support specifying whether the transaction should overwrite, append, or error if the table already exists
+}
+
+#[tokio::main]
+async fn main() -> ExitCode {
+    env_logger::init();
+    match try_main().await {
+        Ok(()) => ExitCode::SUCCESS,
+        Err(e) => {
+            println!("{e:#?}");
+            ExitCode::FAILURE
+        }
+    }
+}
+
+// TODO: Update the example once official write APIs are introduced (issue#1123)
+async fn try_main() -> DeltaResult<()> {
+    let cli = Cli::parse();
+
+    let path = &cli.location_args.path;
+    // TODO: Check if path is a directory and if not, create it
+    let url = delta_kernel::try_parse_uri(path)?;
+    println!("Using Delta table at: {url}");
+
+    // Get the engine
+    let engine = common::get_engine(&url, &cli.location_args)?;
+    let store = engine
+        .get_object_store_for_url(&url)
+        .unwrap_or_else(|| panic!("Failed to get object store for URL: {url}"));
+
+    // Create or get the table
+    let snapshot =
+        Arc::new(create_or_get_base_snapshot(&url, &engine, &cli.write_args.schema, &store).await?);
+
+    // Create sample data based on the schema
+    let sample_data = create_sample_data(&snapshot.schema(), cli.write_args.num_rows)?;
+
+    // Write sample data to the table
+    let mut txn = snapshot
+        .transaction()?
+        .with_operation("INSERT".to_string())
+        .with_engine_info("delta-kernel-rs example");
```

Comment:
can we say something like `"default_engine:v{CRATE_VERSION}/write-table-example"`

Reply:
I assume that you would want to insert `delta_kernel`'s version here? AFAIK, there is no "easy solution" to access the version of dependencies in code as long as the create itself (i.e., delta-kernel) does not expose a constant or macro for it. You could use `cargo_metadata` and a build script, but I would consider this out of scope since it will make the PR even bigger and the example more complex.